storage/storage.go

```diff
@@ -0,0 +1,220 @@
+// Copyright 2016 The OPA Authors.  All rights reserved.
+// Use of this source code is governed by an Apache2
+// license that can be found in the LICENSE file.
+
+package storage
+
+import "github.com/open-policy-agent/opa/ast"
+
+// Config represents the configuration for the policy engine's storage layer.
+type Config struct {
+	Builtin Store
+}
+
+// Storage represents the policy engine's storage layer.
+type Storage struct {
+	builtin Store
+	indices *Indices
+	mounts  []*mount
+}
+
+type mount struct {
+	path    ast.Ref
+	strpath []string
+	backend Store
+}
+
+// New returns a new instance of the policy engine's storage layer.
+func New(config Config) *Storage {
+	return &Storage{
+		builtin: config.Builtin,
+		indices: NewIndices(),
+	}
+}
+
+// Mount adds a store into the storage layer at the given path. If the path
+// conflicts with an existing mount, an error is returned.
+func (s *Storage) Mount(backend Store, path ast.Ref) error {
+	for _, m := range s.mounts {
+		if path.HasPrefix(m.path) || m.path.HasPrefix(path) {
+			return mountConflictError()
+		}
+	}
+	spath := make([]string, len(path))
+	for i, x := range path {
+		switch v := x.Value.(type) {
+		case ast.String:
+			spath[i] = string(v)
+		case ast.Var:
+			spath[i] = string(v)
+		default:
+			return internalError("bad mount path: %v", path)
+		}
+	}
+	m := &mount{
+		path:    path,
+		strpath: spath,
+		backend: backend,
+	}
+	s.mounts = append(s.mounts, m)
+	return nil
+}
+
+// Unmount removes a store from the storage layer. If the path does not locate
+// an existing mount, an error is returned.
+func (s *Storage) Unmount(path ast.Ref) error {
+	for i := range s.mounts {
+		if s.mounts[i].path.Equal(path) {
+			s.mounts = append(s.mounts[:i], s.mounts[i+1:]...)
+			return nil
+		}
+	}
+	return notFoundRefError(path, "unmount")
+}
+
+type hole struct {
+	path []string
+	doc  interface{}
+}
+
+// Read fetches the value in storage referred to by path. The path may refer to
+// multiple stores in which case the storage layer will fetch the values from
+// each store and then stitch together the result.
```

Comment:
Currently, there are rules included in the return value.

Still TBD whether it makes sense to include rules in the return value or expect callers to lookup rules separately (and then eval and stitch the result).